services/sky/document_view.cc

```diff
@@ -99,30 +95,64 @@ base::WeakPtr<DocumentView> DocumentView::GetWeakPtr() {
   return weak_factory_.GetWeakPtr();
 }
 
-void DocumentView::OnEmbed(
-    mojo::View* root,
-    mojo::InterfaceRequest<mojo::ServiceProvider> services_provided_to_embedder,
-    mojo::ServiceProviderPtr services_provided_by_embedder) {
-  root_ = root;
+void DocumentView::InitViewport() {
+  mojo::ServiceProviderPtr viewport_service_provider;
+  shell_->ConnectToApplication("mojo:native_viewport_service",
+                               mojo::GetProxy(&viewport_service_provider),
+                               nullptr);
+  mojo::ConnectToService(viewport_service_provider.get(), &viewport_service_);
+  viewport_service_.set_connection_error_handler(
+      [this]() { OnViewportConnectionError(); });
+
+  mojo::NativeViewportEventDispatcherPtr ptr;
+  event_dispatcher_binding_.Bind(GetProxy(&ptr));
+  viewport_service_->SetEventDispatcher(ptr.Pass());
+
+  // Match the Nexus 5 aspect ratio initially.
+  auto size = mojo::Size::New();
+  size->width = 320;
+  size->height = 640;
+
+  auto requested_configuration = mojo::SurfaceConfiguration::New();
+
+  viewport_service_->Create(size.Clone(),
+                            requested_configuration.Pass(),
+                            base::Bind(&DocumentView::OnViewportCreated,
+                                       base::Unretained(this)));
```

Comment:
This is OK since it's binding a response callback to the lifetime of the object that owns the pipe (in this case |viewport_service_|.  If |this| dies with a response pending it'll tear down |viewport_service_| which will ensure the response callback will never fire.  Of course, this is only true so long as the lifetime of |viewport_service_| remains bounded by the lifetime of the bound target.

also we should be consistent about using base::Bind() or lambdas for this sort of dispatching - I'm fine with either but not both in the same file.